src/operators/products.c

```diff
@@ -0,0 +1,221 @@
+   IGRAPH_SAFE_MULT(ecountg1, ecountg2, &ecount);
+   if (!directed) { // directed tensor product has only e1*e2 edges, see below
+      IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   }
+   IGRAPH_SAFE_MULT(ecount, 2, &ecount);
+   IGRAPH_VECTOR_INT_INIT_FINALLY(&edges, ecount);
+
+   igraph_integer_t from1, to1, from2, to2;
+   igraph_integer_t i, j;
+   
+   // Vertex ((i, j)) with i from g1, and j from g2
+   //   will have new vertex id: i * vcountg2 + j
+   igraph_integer_t edge_index = 0;
+
+   for (i = 0; i < ecountg1; ++i) {
+      from1 = IGRAPH_FROM(g1, i);
+      to1 = IGRAPH_TO(g1, i);
+
+      for (j = 0; j < ecountg2; ++j) {
+         from2 = IGRAPH_FROM(g2, j);
+         to2 = IGRAPH_TO(g2, j);
+
+         // create edge between ((from1, from2)) to ((to1, to2))
+         VECTOR(edges)[edge_index++] = from1 * vcountg2 + from2; // ((from1, from2))
+         VECTOR(edges)[edge_index++] = to1   * vcountg2 + to2;   // ((to1, to2))
+
+         // this cross edge is not present in directed edge
+         // as to2 is not adjacent to from2, if direction is taken in account
+         if (!directed) {
+            VECTOR(edges)[edge_index++] = from1 * vcountg2 + to2;   // ((from1, to2))
+            VECTOR(edges)[edge_index++] = to1   * vcountg2 + from2; // ((to1, from2))
+         }
+      }
+   }
+
+   IGRAPH_CHECK(igraph_create(res, &edges, vcount, directed));
+   igraph_vector_int_destroy(&edges);
+   IGRAPH_FINALLY_CLEAN(1);
+
+   return IGRAPH_SUCCESS;
+}
+
+/**
+ * \function igraph_product
+ * \brief Computes the graph product of two graphs, based on the specified product type.
+ * </para><para>
+ * Supported graph product types are:
+ * \clist
+ *    \cli IGRAPH_PRODUCT_CARTESIAN
+ *       Computes the Cartesian product of two graphs \c g1 and \c g2.
+ * The Cartesian product of two graphs \c g1 and \c g2 is a graph \c res such that:
+ * \olist
+ *    \oli The vertex set of \c res is the Cartesian product of the vertex sets of g1 and g2: V(g1) x V(g2).
+ * </para><para>
+ *    \oli Two vertices <code>(u, v)</code> and <code>(u1, v1)</code> are adjacent in 
+ *    \c res if and only if either <code>u = u1</code> and \c v is adjacent 
+ *    to \c v1 in \c g2, or <code>v = v1</code> and \c u is adjacent to 
+ *    \c u1 in \c g1.
+ * \endolist
+ * Thus, the number of vertices in \c res is |V1| x |V2|, and the number of edges in
+ * \c res is |V1| × |E2| + |V2| × |E1|.
+ * </para><para>
+ * Time Complexity: O(|V1| × |V2| + |V1| × |E2| + |V2| × |E1|)
+ *       where |V1| and |V2| are the number of vertices, and
+ *       |E1| and |E2| are the number of edges in \c g1 and \c g2 respectively.
+ * 
+ * \endclist
+ * Both graphs must be of the same type, either directed or undirected. If a product of an undirected and a
+ * directed graph is required, the undirected graph can be converted to a directed graph using \ref 
+ * igraph_to_directed, or the directed graph can be converted to an undirected graph using \ref 
```

Comment:
Conventionally, when referring to functions, we always add parentheses.

Use `\ref fun()` instead of `\ref fun`.